promql/engine.go

```diff
@@ -1122,6 +1122,13 @@ func (ev *evaluator) rangeEval(prepSeries func(labels.Labels, *EvalSeriesHelper)
 			warnings.Merge(ws)
 			matrixes[i] = val.(Matrix)
 
+			// Since we are copying the samples in memory here, check for ErrTooManySamples.
+			rangeEvalSamples += matrixes[i].TotalSamples()
```

Comment:
> By just adding that size again, aren't we doing the same thing twice?

We are doing that because we are copying the samples in line [1123](https://github.com/prometheus/prometheus/pull/9626/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR1123)

Reply:
note, this check prevents possible OOM error since we check `maxSamples` after loading each expression, and not while processing them post loading...